### DISCUSSION TITLE
Problem using SpanCategorizer - can't initialize labels

### DISCUSSION THREAD

**Discussion by lnatspacy** [explosion/spaCy#8953] [Help: Coding & Implementations] [answered]
Hi, 

I'm trying to use the new SpanCategorizer and I'm having trouble putting things together. 
I annotated 55 samples in Prodigy (I know it's no nearly enough, I just wanted something to test the workflow with ;)).

Here's my config:

```
[paths]
train = null
dev = null
vectors = null
init_tok2vec = null

[system]
gpu_allocator = null
seed = 0

[nlp]
lang = "de"
pipeline = ["spancat"]
batch_size = 1000
disabled = []
before_creation = null
after_creation = null
after_pipeline_creation = null
tokenizer = {"@tokenizers":"spacy.Tokenizer.v1"}

[components]

[components.spancat]
factory = "spancat"
max_positive = null
spans_key = "grit"
threshold = 0.5

[components.spancat.model]
@architectures = "spacy.SpanCategorizer.v1"

[components.spancat.model.reducer]
@layers = "spacy.mean_max_reducer.v1"
hidden_size = 128

[components.spancat.model.scorer]
@layers = "spacy.LinearLogistic.v1"
nO = null
nI = null

[components.spancat.model.tok2vec]
@architectures = "spacy.Tok2Vec.v1"

[components.spancat.model.tok2vec.embed]
@architectures = "spacy.MultiHashEmbed.v1"
width = 96
rows = [5000,2000,1000,1000]
attrs = ["ORTH","PREFIX","SUFFIX","SHAPE"]
include_static_vectors = false

[components.spancat.model.tok2vec.encode]
@architectures = "spacy.MaxoutWindowEncoder.v1"
width = 96
window_size = 1
maxout_pieces = 3
depth = 4

[components.spancat.suggester]
@misc = "spacy.ngram_range_suggester.v1"
min_size = 1
max_size = 40

[corpora]

[corpora.dev]
@readers = "spacy.Corpus.v1"
path = ${paths.dev}
max_length = 0
gold_preproc = false
limit = 0
augmenter = null

[corpora.train]
@readers = "spacy.Corpus.v1"
path = ${paths.train}
max_length = 2000
gold_preproc = false
limit = 0
augmenter = null

[training]
dev_corpus = "corpora.dev"
train_corpus = "corpora.train"
seed = ${system.seed}
gpu_allocator = ${system.gpu_allocator}
dropout = 0.1
accumulate_gradient = 1
patience = 1600
max_epochs = 0
max_steps = 20000
eval_frequency = 200
frozen_components = []
before_to_disk = null

[training.batcher]
@batchers = "spacy.batch_by_words.v1"
discard_oversize = false
tolerance = 0.2
get_length = null
size = 100

[training.logger]
@loggers = "spacy.ConsoleLogger.v1"
progress_bar = false

[training.optimizer]
@optimizers = "Adam.v1"
beta1 = 0.9
beta2 = 0.999
L2_is_weight_decay = true
L2 = 0.0
grad_clip = 1.0
use_averages = false
eps = 0.00000001
learn_rate = 0.001

[training.score_weights]
spans_grit_f = 1.0
spans_grit_p = 0.0
spans_grit_r = 0.0
spans_sc_f = null
spans_sc_p = null
spans_sc_r = null

[pretraining]

[initialize]
vectors = null
init_tok2vec = ${paths.init_tok2vec}
vocab_data = null
lookups = null
before_init = {"@callbacks":"customize_tokenizer"}
after_init = null

[initialize.components]

[initialize.tokenizer]
```

When I run spacy train I get the following error: 

```
ValueError: [E143] Labels for component 'spancat' not initialized. This can be fixed by calling add_label, or by providi
ng a representative batch of examples to the component's `initialize` method.
```

So I tried to create a label file but when I run spacy init labels, I get the exact same error. 

What am I missing here?

Any help would be much appreciated!

**Comment by adrianeboyd** [comment, marked as answer]
It sounds like you don't have the annotated spans saved in the right place in your training docs. The `SpanCategorizer` trains from the annotation saved in `doc.spans[spans_key]`, not from `doc.ents`. Double-check that you've put the annotation in `doc.spans` under the right `spans_key`, which by default is `sc`, but is `grit` in the config above. If you've annotated NER data in prodigy, see the convert script in that example project that copies `doc.ents` to `doc.spans` to create the training data. Otherwise, you may just need to adjust `spans_key` in your config to the right value. It looks like the new prodigy `spans` recipes use the `sc` default, so maybe just changing that in your config above will work?

**Reply by lnatspacy** [reply]
Sadly, just changing the key to sc didn't work. I used the prodigy recipe spans.manual. I'd have expected that when I run data-to-spacy on such a dataset prodigy would put everything where it needs to be itself. I'll try to run the convert script. 

Thank you! :) 

/edit: 

Thank you, that did the trick ! 👍

**Reply by adrianeboyd** [reply]
Glad to hear it's working!

The default workflow in prodigy with `data-to-spacy` should hopefully just work (I'm not sure how you ended up using this `grit` config?), but if you run into issues that seem more directly related to prodigy, then the prodigy support forum is a good place to post a bug report.

**Reply by polm** [reply]
@vaibhav-01 It sounds like your issue is only vaguely related to this one. Please open a new discussion, post your code, and don't @ individual people.
